### PR TITLE
feat(types): package scaffolding for generated types

### PR DESCRIPTION
**Pull request by patrick-medusajs**
## What

Generated client types package based on the API's OpenApi Spec.

## Why

This package aims at replacing the need to import types directly from the core `@medusajs/medusa` package which can lead to side effects like importing backend dependencies into the client's realm.

## How

* The package contains a script that is invoked during the build step. 
* The script will use the `medusa-oas oas --type=combined` cli to generate a combined oas file. 
* Then, `medusa-oas client --component=types` uses the oas file to generate the types based on the OAS schema definitions. The result is outputted into `/src/lib`
* `prettier` is run on the `/src/lib` directory to ensure that the generated code follows the monorepo coding style.
* Finally, with `"emitDeclarationOnly": true`,  `tsc` compiles the src directory to only output types.

Until we update our CI pipeline, we will not commit the content `/src/lib`.

## Test

* Run `yarn install`
* Run `yarn build`
* Expect `packages/generated/client-types/dist/` to contain types generated from the OAS defined by `@medusajs/medusa`

### CODE REVIEW

**Review by shahednasser — APPROVED** [review]
LGTM - no change in the documentation

File: packages/generated/medusa-client-types/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "@medusajs/medusa-client-types",
```
Comment (olivermrbl) [inline packages/generated/medusa-client-types/package.json:2]: very nit(non-blocking): 
```suggestion
  "name": "@medusajs/client-types",
```

**Review by olivermrbl — APPROVED** [review]
Very nice - tested locally and works as expect ✅

File: packages/generated/medusa-client-types/package.json (hunk 1 above)
Comment (patrick-medusajs) [reply packages/generated/medusa-client-types/package.json]: Good idea, done in 5ab163b2640689e323c9b33c7827a577c5272321

**Pull request merged**